### PR TITLE
add docker, ipm, fix issues 1

### PR DESCRIPTION
**Pull request by SergeyMi37**

### CODE REVIEW

File: source/cache/REST/Informer.cls [hunk 1]
```diff
@@ -47,6 +47,7 @@ ClassMethod GetMemberData(class As %Dictionary.ClassDefinition, memberName As %S
         set p = $PROPERTY(class, propName).GetAt(i)
         for j=1:1:props.Properties.Count() {
             set pname = props.Properties.GetAt(j).Name
+	    try {
             set val = $PROPERTY(p, pname)
```
Comment (nikitaeverywhere) [inline source/cache/REST/Informer.cls]: Fix indentation

File: build/cls/VisualEditor/Installer.cls [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+ClassMethod Init()
+{
+    set installLog = ""
+        set errorOccurred = 0
+        }
+        
+        /// This method creates a new web application by the given spec.
+        ClassMethod RegisterWebApplication(name As %String, spec) As %Status
+        {
+            $$$log("Moving to %SYS.")
```
Comment (nikitaeverywhere) [inline build/cls/VisualEditor/Installer.cls]: Why is it indented like this? Does gulp really build it like this?

**Review by nikitaeverywhere — COMMENTED** [review]
Thanks for your contribution! I have no opportunity to test it right now, can you provide which IRIS version this is tested on? Just for the note purpose.

I assume the build is fully done with `npm run gulp` command.

File: source/cache/REST/Informer.cls [hunk 3]
```diff
@@ -57,6 +58,7 @@ ClassMethod GetMemberData(class As %Dictionary.ClassDefinition, memberName As %S
                 }
                 do op.%Set(pname, v)
             }
+            } catch e {}
```
Comment (nikitaeverywhere) [inline source/cache/REST/Informer.cls]: I'd log the error at least.

**Comment by SergeyMi37** [comment]
Unfortunately, I didn't build with npm run gulp. Indentation in methods doesn't matter as they are overridden by loading from xml

**Comment by nikitaeverywhere** [comment]
> Unfortunately, I didn't build with npm run gulp.

So how did you get that code? You should obtain build artefacts from sources of this repo, otherwise we need to remove sources...

**Comment by SergeyMi37** [comment]
tested on versions:
IRIS for Windows (x86-64) 2022.1.1 
IRIS for Ubuntu (x86-64) 2023.1

**Comment by nikitaeverywhere** [comment]
That's fine, but now we have 2 code copies in different places. This is not good. Either one code must be generated from the source code (build with gulp -> output to cls -> copy to whatever), or we need to remove the source code and keep classes as you suggest in another directory for CPM. But this is not good either, since the repo contains the build pipeline for front end code too.

Can you try running gulp and then if it works, moving files around?

**Comment by SergeyMi37** [comment]
I can't work with gulp.

**Comment by nikitaeverywhere** [comment]
> I can't work with gulp.

Then let me help you a bit (though the build systems deprecated much, it's still runnable):

```bash
$ git clone --recurse-submodules git@github.com:intersystems-community/objectscript-visual-editor.git
...
$ cd objectscript-visual-editor
$ python --version
Python 2.7.18
$ nvm use 10
Now using node v10.24.1 (npm v6.14.12)
$ npm install
...
$ npm run gulp
...
```

See [nvm](https://github.com/nvm-sh/nvm) (available in [brew](https://formulae.brew.sh/formula/nvm)) and how to install python2 ([on mac](https://stackoverflow.com/a/71620699/2590150)) if needed. This is applicable to all my projects.

As a result, you'll get a nice `build` directory:

image

You can extend this build pipeline, like copy files to `build/cls` instead of `build/cache`. As [this thing](https://github.com/intersystems-community/objectscript-visual-editor/pull/2/files#diff-4800e71f9121909cee15bb7ec18eb8f9b0ba40e3ce856deb97b80d6d6352863cR28) you commit won't even work (how did you test it?):

```
XData acejs [ MimeType = text/javascript ]
{
<d><![CDATA[<!-- @include ../client/js/lib/ace/ace.js -->]]></d>
}
```

**Comment by nikitaeverywhere** [comment]
Simply put,

- I suggest you to just rename `cache` -> `cls` (please tell why you need this), without any changes.
- Try running the `gulp` pipeline as described above, and fix the destination dir name (should be just 1 config line).
- Add CPM definitions.
- Done!

**Comment by SergeyMi37** [comment]
Hi, Nikita.
Thanks for the help gulp.
But my goal is not to learn how to continue to develop your project, but to make it work in the ZPM with the least amount of effort. And not to harm the future developer of this package.
To avoid code duplication, I'll rename /cache to /VisualEditor.

**Comment by nikitaeverywhere** [comment]
I don't understand you, and I have doubts you understand me.

The code you submit won't work with IPM, because you simply have these lines

```
XData acejs [ MimeType = text/javascript ]
{
<d><![CDATA[<!-- @include ../client/js/lib/ace/ace.js -->]]></d>
}
```

...in files "you have just renamed", which will serve nothing to the front end (literally, it will serve `<!-- @include ..` line comment), and the web app won't launch.

And I am suggesting you the 

>  least amount of effort

...way to achieve what you want.

1. Build the repo and get the `/build` directory with properly built classes (I can push it if you want; I've just built it locally this morning for you)
2. Add ipm definitions
3. Done!

**Comment by SergeyMi37** [comment]
The code sent to the ZPM will work, as it will be taken from the release:
``` ...
     <FileCopy Name="build/" Target="${mgrdir}.modules/VisualEditor/"/>
      <Resource Name="VisualEditor.REST.Editor.cls" />
      <Invoke Class="%SYSTEM.OBJ" Method="Load" CheckStatus="true">
       <Arg>${mgrdir}.modules/VisualEditor/VisualEditor-v0.10.1.xml</Arg>
       <Arg>ck</Arg>
      </Invoke>
 ...
 ```
This is a similar action
Execute  in terminal.
```
 do $system.OBJ.Load("..\mgrdir\.modules/VisualEditor/VisualEditor-v0.10.1.xml","ck") 
 ```

**Comment by SergeyMi37** [comment]
Of course, this is a hack for the ZPM, but it works.

**Comment by nikitaeverywhere** [comment]
Then why you need to touch cls? (if you only use well-exported *.xml)

image

**Comment by nikitaeverywhere** [comment]
I suggest I push `/build` from my local machine if you need those (for any unknown weird and ugly hack), and then you add XML and IPM definitions.

Or, please remove those cls files. No source code should be copied (while xml can be considered as a build artefact stored within the repo)

**Comment by SergeyMi37** [comment]
The specificity of the ZPM is that you need to specify resources in a specific format. The resource name format includes a fully qualified or unqualified package name, or a fully qualified class name.

**Comment by nikitaeverywhere** [comment]
That's OK. What should we do?

**Comment by SergeyMi37** [comment]
Do it well.

**Comment by nikitaeverywhere** [comment]
Discussed over the phone. I've uploaded built classes, you can now rebase and add just the IPM definitions into this PR.

image

**Comment by SergeyMi37** [comment]
There's also a cant implementation of MimeType in XData
#12 55.34 [objectscript-visual-editor]  Compile FAILURE
#12 55.34 ERROR! The class definition for class 'VisualEditor.REST.Router' could not be parsed correctly, possibly due to non-matching {} or () characters or a missing ; character or non-matched /* */ blocks. Edit the class and correct the problem.
#12 55.34   > ERROR #5030: An error occurred while compiling class 'VisualEditor.REST.Router'
image
To fix this, you need MimeType = application/yaml
image

**Comment by SergeyMi37** [comment]
There's also a cant implementation of MimeType in XData
Errors are generated during compilation.
```
#12 24.40 [USER|objectscript-visual-editor]     Reload SUCCESS
#12 24.40 [objectscript-visual-editor]  Module object refreshed.
#12 24.40 [USER|objectscript-visual-editor]     Validate START
#12 24.40 [USER|objectscript-visual-editor]     Validate SUCCESS
#12 24.40 [USER|objectscript-visual-editor]     Compile START
#12 24.62 Compilation started on 06/22/2023 13:00:41 with qualifiers 'd-lck'
#12 24.64 Compiling 6 classes
#12 24.66 ERROR #5559: The class definition for class 'VisualEditor.REST.Router' could not be parsed correctly, possibly due to non-matching {} or () characters or a missing ; character or non-matched /* */ blocks. Edit the class and correct the problem.
#12 24.66   > ERROR #5030: An error occurred while compiling class 'VisualEditor.REST.Router'
#12 24.95 Compiling class VisualEditor.REST.Editor
#12 24.96 Compiling class VisualEditor.REST.Informer
#12 24.97 Compiling class VisualEditor.REST.Content
#12 24.98 Compiling class VisualEditor.REST.StaticLibJS
#12 24.99 Compiling class VisualEditor.Installer
#12 25.02 Compiling routine VisualEditor.REST.Content.1
#12 25.05 Compiling routine VisualEditor.REST.Editor.1
#12 25.06 Compiling routine VisualEditor.REST.Informer.1
#12 25.06 Compiling routine VisualEditor.REST.StaticLibJS.1
#12 25.08 Compiling routine VisualEditor.Installer.1
#12 55.34 Installing Cache Visual Editor to USER
#12 55.34 Moving to %SYS.
#12 55.34 Creating WEB application "/VisualEditor"...
#12 55.34 WEB application "/VisualEditor" is created.
#12 55.34 Installation is complete!
#12 55.34 Detected 1 errors during compilation in 30.72s.
#12 55.34
#12 55.34 [objectscript-visual-editor]  Compile FAILURE
#12 55.34 ERROR! The class definition for class 'VisualEditor.REST.Router' could not be parsed correctly, possibly due to non-matching {} or () characters or a missing ; character or non-matched /* */ blocks. Edit the class and correct the problem.
#12 55.34   > ERROR #5030: An error occurred while compiling class 'VisualEditor.REST.Router'
```
Made through installation from xml

**Comment by nikitaeverywhere** [comment]
> Made through installation from xml

Maybe cls? Wondering what could be the fix

**Comment by SergeyMi37** [comment]
I do not know what to do.
1. The method of loading from XML
```
      <SourcesRoot>build</SourcesRoot>
      <FileCopy Name="build/" Target="${mgrdir}.modules/VisualEditor/"/>
      <Resource Name="VisualEditor.REST.Editor.cls" />
      <Invoke Class="%SYSTEM.OBJ" Method="Load" CheckStatus="true">
       <Arg>${mgrdir}.modules/VisualEditor/VisualEditor-v0.10.1.xml</Arg>
       <Arg>ck</Arg>
      </Invoke>
```
 is universal, it works on all the latest versions of IRIS

2. Download method from cls files 
```
      <SourcesRoot>build</SourcesRoot>
      <Resource Name="VisualEditor.PKG" />
      <Invoke Class="VisualEditor.Installer" Method="CreateProjection" CheckStatus="true">
      </Invoke>
```
on versions higher than intersystemsdc/iris-community:2022.1.1.374.0-zpm
gives an error message
```
#12 24.66 ERROR #5559: The class definition for class 'VisualEditor.REST.Router' could not be parsed correctly, possibly due to non-matching {} or () characters or a missing ; character or non-matched /* */ blocks. Edit the class and correct the problem.
#12 24.66   > ERROR #5030: An error occurred while compiling class 'VisualEditor.REST.Router'
```
Apparently you need to start an appeal to the WRC

**Comment by nikitaeverywhere** [comment]
> is universal, it works on all the latest versions of IRIS

I recall it works starting from 2018 or so, and actually VisualEditor was written specifically for these "new" versions of former Cache (IRIS), because it uses native JSON under the hood. So we can forget about XML and import CLS straight away.

If it doesn't work after whatever you fix in CLS, feel free to make this PR contain that XML to import, and "reference" a non-erroring CLS class, this is fine (right now, the PR still contains CLS copies, please remove it).

**Comment by SergeyMi37** [comment]
Import problem from cls format in this section of code
image
class VisualEditor.REST.Router

**Comment by nikitaeverywhere** [comment]
Got it. I think InterSystems broke backwards compatibility again. The code uses MimeType [here](https://github.com/intersystems-community/objectscript-visual-editor/blob/5880b807d455812ebe175aac29bd1ddfd4c4d6a9/source/cache/REST/Content.cls#L17) to serve the correct type to the front end.

I suggest we keep your XML file as a main thing to import. And use that "quick fix" with CLS by linking just one non-conflicting file.

image

**Comment by SergeyMi37** [comment]
Fulfilled

**Pull request merged**